### DISCUSSION TITLE
New API method does nothing

### DISCUSSION THREAD

**Discussion by PeterParkerLogic** [microsoft/AirSim#3253] [Support Q&A] [answered]
Is there any method to let drone can fly forward ,backward, left turn and right turn in unreal engine by using python ?
I want to let the drone fly forward to the direction what the center camera see. Not fly with world (NED) axis. I also try to use moveByVelocityBodyFrameAsync(), but the drone in unreal engine does not do anything by using this api.

thank you in advance.

**Comment by ahmed-elsaharti** [comment]
Hi @PeterParkerLogic, you're right, the API that does what you're trying to do should be the `moveByVelocityBodyFrameAsync` API. When you try to use it, what are the input arguments are you passing to the `moveByVelocityBodyFrameAsync` function? Also, which version of Airsim are you testing this with (ie: is it a clone of the latest repo or is it one of the precompiled environments in the releases) ?

**Reply by PeterParkerLogic** [reply]
You can see the video(https://drive.google.com/file/d/1BjocfQspBolaKEId0Xua5PKy3PaF7Mpf/view?usp=sharing) that moveByVelocityBodyFrameAsync API could not work. I upgraded my airsim to latest version yesterday and rebuild the airsim by `./build.sh`. This is my code. 
`"""
For connecting to the AirSim drone environment and testing API functionality
"""
import setup_path 
import airsim

import os
import tempfile
import pprint

from imutils import perspective
from imutils import contours
import numpy as np
import imutils
import cv2
from shapely.geometry import Point
from shapely.geometry.polygon import Polygon
from random import randrange

#connect to the AirSim simulator
client = airsim.MultirotorClient()
client.confirmConnection()
client.enableApiControl(True)
client.armDisarm(True)

state = client.getMultirotorState()
s = pprint.pformat(state)
print("state: %s" % s)



landed = client.getMultirotorState().landed_state
if landed == airsim.LandedState.Landed:
    print("taking off...")
    client.takeoffAsync().join()
else:
    print("already flying...")
    client.hoverAsync().join()

client.moveByManualAsync(vx_max = 1E6, vy_max = 1E6, z_min = -1E6, duration = 9E10)

import sys, termios, tty, os, time

#start go up
vehicleControl = client.moveByVelocityAsync(0, 0, -1, 2)
vehicleControl.join()
time.sleep(5)
print("start")
vehicleControl = client.moveByVelocityBodyFrameAsync(5, 0, 0, 3)
vehicleControl.join()
print("over")
time.sleep(5)

landed = client.getMultirotorState().landed_state
if landed == airsim.LandedState.Landed:
    print("already landed...")
else:
    print("landing...")
    client.landAsync().join()

client.armDisarm(False)

client.enableApiControl(False)`
Thank you

**Reply by ahmed-elsaharti** [reply]
Hi @PeterParkerLogic , I've tested your code and it works perfectly on my machine (although I had to get rid of the `imutils` and `tty` importation because I don't have those modules installed).
Just to verify, where is your python script located in relation to the repo's directory? Also, does the execution of the python script return any errors?

**Reply by PeterParkerLogic** [reply]
My program is located in Airsim/PythonClient/multirotors. There was no error when I executed this code. Maybe I should reload airsim to unreal engine project?  Because I only rebuilded latest airsim , I did not change anything in my unreal project.

**Reply by jonyMarino** [reply]
Hi! Yes. You can copy the scripts from the Blocks project and run update_from_git. This is necessary to run a new API function.

**Reply by ahmed-elsaharti** [reply]
@PeterParkerLogic yup as @jonyMarino mentioned, since you seem to be using a custom environment, you'll need to rebuild the repo and copy the plugin over again like shown [here](https://microsoft.github.io/AirSim/unreal_custenv/#step-by-step-instructions). This updates AirSim's files within your UE project to reflect the addition of the new API (this is assuming your current environment has been built a while ago before the addition of that API)

**Reply by PeterParkerLogic** [reply]
Thank you very much. It works for me.

**Comment by jonyMarino** [comment, marked as answer]
> PeterParkerLogic yup as jonyMarino mentioned, since you seem to be using a custom environment, you'll need to rebuild the repo and copy the plugin over again like shown [here](https://microsoft.github.io/AirSim/unreal_custenv/#step-by-step-instructions). This updates AirSim's files within your UE project to reflect the addition of the new API (this is assuming your current environment has been built a while ago before the addition of that API)

**Reply by jonyMarino** [reply]
Thanks @ahmed-elsaharti !
